lib/api/client-commands/setGeolocation.js

```diff
@@ -0,0 +1,77 @@
+const ClientCommand = require('./_base-command.js');
+const {Logger} = require('../../utils');
+
+/**
+ * Mock the geolocation of the browser.
+ *
+ * @example
+ *  this.demoTest = function (browser) {
+ *    browser
+ *      // Set location of Tokyo, Japan
+ *      .setGeolocation({
+ *        latitude: 35.689487,
+ *        longitude: 139.691706,
+ *        accuracy: 100
+ *        })
+ *      .navigateTo('https://kawasaki-india.com/dealer-locator/');
+ *  };
+ *
+ * @method setGeolocation
+ * @syntax .setGeolocation({latitude, longitude, accuracy}, [callback])
+ * @param {object} coordinates
+ * @param {function} [callback] Optional callback function to be called when the command finishes.
+ * @api protocol.userprompts
+ */
+class SetGeolocation extends ClientCommand {
+
+  performAction(callback) {
+
+    if (!this.api.isChrome()  && !this.api.isEdge()) {
+      const error =  new Error('SetGeolocation is not supported while using this driver');
+      Logger.error(error);
+
+      return callback(error);
+    }
+
+    const {latitude, longitude, accuracy} = this.coordinates;
+
+    if (latitude === undefined && longitude === undefined) {
+      // Clear geolocation override.
+      this.transportActions
+        .clearGeolocation(callback)
+        .catch(err => {
+          return err;
+        })
+        .then(result => callback(result));
+
+    } else if (latitude === undefined || longitude === undefined) {
+      // Exactly one of them is undefined, throw error.
+      const error =  new Error('Please provide both latitude and longitude while using setGeolocation.');
+      Logger.error(error);
+
+      return callback(error);
+
+    } else {
```

Comment:
else after return is not needed.

Reply:
But there is an `if` condition before that without any return statement, so removing `else` here would result in this block running even when the condition in `if` is true.

Also, I think it's more clear this way what we are trying to do, as all the three conditions relate to each other: if (both are undefined) -- else if (either of them is undefined) -- else (none of them is undefined).

Reply:
Yes, that needs to be refactored as well. There's a large bit of code being duplicated.

Reply:
Is this fine? I can't think of any other way to refactor it further.
```js
    const {latitude, longitude, accuracy} = this.coordinates;

    if (latitude !== undefined && longitude !== undefined) {
      // Both latitude and longitude are provided

      const coordinates = {latitude, longitude, accuracy};
      // Set accuracy as 100 if not provided.
      if (accuracy === undefined) {coordinates.accuracy = 100}

      this.transportActions
        .setGeolocation(coordinates, callback)
        .catch(err => {
          return err;
        })
        .then(result => callback(result));

      return;
    }

    if (latitude === undefined && longitude === undefined) {
      // Clear geolocation override.
      this.transportActions
        .clearGeolocation(callback)
        .catch(err => {
          return err;
        })
        .then(result => callback(result));

      return;
    }

    // Exactly one of them is undefined, throw error.
    const error =  new Error('Please provide both latitude and longitude while using setGeolocation.');
    Logger.error(error);

    return callback(error);
```

Reply:
`callback` is called twice.

Reply:
I've been facing this issue for many days now, but finally figured it out. 

It's happening because if the command runs correctly without any error, the `callback` is called twice, once in the last line of `transportActions()` method in `commands.js` and then in the `then` clause above. But if some error occurs, the `callback` is just called once in the `then` clause above (because there's no `catch` clause in the last line of `transportActions()`.

So, this issue can be solved if we just call the `callback` above in case some error occurs, i.e., in the `catch` clause:
```js
this.transportActions
  .setGeolocation(coordinates, callback)
  .catch(err => {
    Logger.error(err);  // to log the err in the terminal also, which is not done now
    callback(err);
  });
```

Reply:
Fixed `callback` being called twice issue in 0f67a80 and also tested the commands, both with and without error.

Reply:
Refactored the `setGeolocation` command in [392a0bc](https://github.com/nightwatchjs/nightwatch/pull/3208/commits/392a0bc68563adf0d3e713b58fec50a0778944d0).